File: Array-String/HashTable/JS/degree_of_an_array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findShortestSubArray = function(nums) {
    const freq = new Map();
    const firstIdx = new Map();
    const lastIdx = new Map();

    for (let i = 0; i < nums.length; ++i) {
        const x = nums[i];
        freq.set(x, (freq.get(x) || 0) + 1);
        if (!firstIdx.has(x)) firstIdx.set(x, i);
        lastIdx.set(x, i);
    }

    let degree = 0, minLen = nums.length;

    for (const [num, count] of freq.entries()) {
        if (count > degree) {
            degree = count;
            minLen = lastIdx.get(num) - firstIdx.get(num) + 1;
        } else if (count == degree) {
            minLen = Math.min(minLen, lastIdx.get(num) - firstIdx.get(num) + 1);
        }
    }

    return minLen;
};

File: Array-String/HashTable/Python/degree_of_an_array.py
class Solution:
    def findShortestSubArray(self, nums: list[int]) -> int:
        freq, first, last = {}, {}, {}
        for i, n in enumerate(nums):
            freq[n] = freq.get(n, 0) + 1
            if n not in first: first[n] = i
            last[n] = i
        minLen, degree = len(nums), 0
        for n, count in freq.items():
            if count > degree:
                degree = count
                minLen = last[n] - first[n] + 1
            elif count == degree:
                minLen = min(last[n] - first[n] + 1, minLen)
        return minLen